runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineResult.java

```diff
@@ -178,19 +182,36 @@ protected void stop() {
       // after calling stop, if exception occurs in "grace period" it won't propagate.
       // calling the StreamingContext's waiter with 0 msec will throw any error that might have
       // been thrown during the "grace period".
-      javaStreamingContext.awaitTermination(0);
-      SparkContextFactory.stopSparkContext(javaSparkContext);
+      try {
+        javaStreamingContext.awaitTermination(0);
+      } catch (Exception e) {
+        throw beamExceptionFrom(e);
+      } finally {
+        SparkContextFactory.stopSparkContext(javaSparkContext);
+        if (Objects.equals(state, State.RUNNING)) {
+          state = State.STOPPED;
+        }
+      }
     }
 
     @Override
-    protected State awaitTermination(final Duration duration) throws TimeoutException,
-        ExecutionException, InterruptedException {
-      pipelineExecution.get(duration.getMillis(), TimeUnit.MILLISECONDS);
-      if (javaStreamingContext.awaitTerminationOrTimeout(duration.getMillis())) {
-        return State.DONE;
-      } else {
-        return null;
-      }
+    protected State awaitTermination(final Duration duration) throws ExecutionException,
+        InterruptedException {
+      pipelineExecution.get(); // execution is asynchronous anyway so no need to time-out.
+      javaStreamingContext.awaitTerminationOrTimeout(duration.getMillis());
+
+      switch (javaStreamingContext.getState()) {
+         case ACTIVE:
+           this.state = State.RUNNING;
+           break;
+         case STOPPED:
+           this.state = State.DONE;
+           break;
+         default:
+           this.state = null;
```

Comment:
I think we can avoid mutating `state` here, and just return the appropriate `State` back to the caller.
Since `SparkPipelineResult#waitUntilFinish()` stores the `state` returned by `awaitTermination(...)` anyway, the bottom line should be the same with the advantage we'll only have one place where `state` is set.

Also perhaps if the state is neither `DONE` nor `RUNNING`, then `UNKNOWN` should be returned so that users are not taken by surprise getting a `null` here.